### PR TITLE
Assorted improvements to the Battle.net backend

### PR DESCRIPTION
**Pull request by jleclanche**
Alrighty, now that 8dc2f2d5cc3ce0e5e1b999129ceaa57ed4e75390 has landed I can go ahead with this :)

The first commit starts addressing #1442. It doesn't outright *remove* the adapter but it makes it way less nasty. The current adapter *always* causes the "user", "user1"... mechanism because the battletag is never saved. Additionally, the current adapter always forces at least two sql writes (first the user, then re-saves the user's battletag).

So we first allow overriding the `username_regex` attribute. Rather than make it a setting, we allow specifying it in the adapter, which I believe to be a reasonable solution.
With that done, we can change the Battle.net adapter to use that and get the battletag as the username data.

The next commits address the china issue:

- `valid_region` is simply a list of the valid region inputs for `battlenet_region`
- Appending `-cn` to the UIDs matching `cn` region is required because the UIDs can clash between US and China; if we don't do that, an account in china can take over an account in the US if running both on the same server.
- Finally, the last commit allows overriding the region in `/account/battlenet/login/` by specifying, for example, `?region=cn`. I'm not sure if there is existing functionality for this but I think it's acceptable.

### CODE REVIEW

**Comment by jleclanche** [comment]
@pennersr please review :)

**Comment by jleclanche** [comment]
Do not merge

**Comment by jleclanche** [comment]
@pennersr So the obvious problem I overlooked when specifying `/account/battlenet/login/?region=cn` we lose it after we return from upstream, ending up in `/account/battlenet/login/callback/?code=x&state=y` but missing `&region=cn`.

So I think I need to be able to specify the callback url. I see that the only way to do that right now is to override `OAuth2View.get_client()`, which sounds overzealous although a couple of providers do do that.

What do you think of adding a method `get_callback_url(self, request)` to OAuth2View?

**Comment by jleclanche** [comment]
Disregard the last message. I added that method anyway because it's useful but it's not necessary for the rest of the commits.

Everything works right now. I have to wait for tomorrow for somebody to test in china, though.

I'm not a *huge* fan of how much code is involved in passing the region around, but I really do think it's the correct solution.

File: allauth/socialaccount/providers/battlenet/provider.py [hunk 1]
```diff
@@ -1,18 +1,12 @@
...
 class BattleNetSocialAccountAdapter(DefaultSocialAccountAdapter):
-    def save_user(self, request, sociallogin, form=None):
-        user = super().save_user(request, sociallogin, form)
-
-        battletag = sociallogin.account.extra_data.get("battletag")
-        if battletag:
-            user.username = battletag
-            user.save()
-        return user
+    username_regex = re.compile(r"^[\w.]+#\d+$")
```
Comment (pennersr) [inline allauth/socialaccount/providers/battlenet/provider.py]: If BattleNet requires this particular regexp, why not simply add it to the documentation. You could even go as far doing this:

```
from allauth.account.adapter import get_adapter

if get_adapter().uername_regex != 'this-one':
  raise ImproperlyConfigured()

```

Then, we the issue remains contained within this specific adapter/provider, and developers won't be able to accidentally wrongly configure things.

File: allauth/socialaccount/providers/oauth2/views.py [hunk 2]
```diff
@@ -64,12 +64,17 @@ def view(request, *args, **kwargs):
                 return e.response
         return view
 
-    def get_client(self, request, app):
+    def get_callback_url(self, request, app):
```
Comment (pennersr) [inline allauth/socialaccount/providers/oauth2/views.py]: Wouldn't the OAuth2Adapter be a better place to move this into? You already have a BattleNetOAuth2Adapter ...

Comment (jleclanche) [inline allauth/socialaccount/providers/oauth2/views.py]: This particular change is not related to the BattleNetOAuth2Adapter - see my comments above. I left it in because I think it improves the code.

**Comment by jleclanche** [comment]
@pennersr Replying to the aggregate comments regarding `username_regex`:

I agree that it doesn't ideally belong in the social adapter. However, we then need to figure out what to do for Battle.net by default. I think it's very user unfriendly to require writing a bunch of extra code in order to use the adapter - by removing the adapter, that would be what's needed.

There is one alternative: We drop the `#`. So by default, the username would be, instead of `adys#1234`, just `adys`. If there's already an `adys` in the system, I would *like* to pass the `1234` down to the unique username generator as a hint, so that we end up with `adys_1234` or `adys1234`, something like that. Or, we can always replace it with a dot, as the dot is not used upstream at all.

I'm not a fan of any of this - I have gotten to really like having the full Battle.net username available, battletag included. But maybe that's just me having gotten used to it. I don't mind, I can always override the adapter for my own project.

Either way, I need you to take a decision on this: Do we...
- Continue trying to find a way to have the hash by default?
- Replace the hash with a dot/dash/underscore?
- Strip it and fall back to the generator (and if so how?)?
- Accept `username_regex` as a valid feature and land this as is?

Again, keep in mind that whatever the outcome of this is, Discord is using a similar system for usernames so we have to find something that works for both (I intend to commit a Discord backend at some point).

File: allauth/socialaccount/helpers.py [hunk 3]
```diff
@@ -20,40 +20,39 @@
...
-            username = user_username(sociallogin.user)
-            try:
-                get_account_adapter().clean_username(username)
-            except ValidationError:
-                # This username is no good ...
-                user_username(sociallogin.user, '')
-        # FIXME: This part contains a lot of duplication of logic
-        # ("closed" rendering, create user, send email, in active
-        # etc..)
+        return HttpResponseRedirect(url)
```
Comment (jleclanche) [inline allauth/socialaccount/helpers.py]: Can I introduce you to the wonderful, wonderful `q` library? :)

https://pypi.python.org/pypi/q

Comment (jleclanche) [inline allauth/socialaccount/helpers.py]: I find this style a ton cleaner and more readable. I can change it back if you insist, but I'm putting up a little fight ;)

**Comment by pennersr** [comment]
How about the following approach.

- I'll prep a change where the username regex becomes pluggable, based on this: https://github.com/django/django/blob/1.10/django/contrib/auth/validators.py

- Then, you can provide a BattleNetUsernameValidator in your provider. Additionally, you can do an assert inside the constructor of your adapter that a battlnet username actually matches the currently configured username regex.

People have been asking for a configurable regex anyway, so this tackles 2 issues in one shot..

**Comment by jleclanche** [comment]
Awesome! Yes, that sounds great.

**Comment by pennersr** [comment]
The pluggable regex is in, see: 4e534e0e1f39753bc3001c1d3427757d084db90f

**Comment by pennersr** [comment]
I am still pondering about the regex now that it is in. Basically, it does not make much sense to have a setting for that, as under normal circumstances people that have special requirements should have setup a custom user model anyway. There, one is free to setup whatever username validator.

The problem I see is that if you ignore the above and use a configurable validator that is solely used by allauth, then you will run into trouble saving your user instance from the Django admin later on.

**Comment by jleclanche** [comment]
@pennersr Just let me know how you'd like to see it laid out and I'll push a commit out today/tomorrow.

**Comment by pennersr** [comment]
Are you using a custom user model? If so, then I recommend you to put the proper username validator in there.

**Comment by pennersr** [comment]
As for this PR, I would:

- Undo the username_regex changes from the social adapter

- Add a BattlenetUsernameValidator

- Add a note in the docs that this validator should be used by setting up a custom user model using: `username = models.CharField(validators=[BattlenetUsernameValidator(),..])` 

- Undo the `ret` related changes (sorry for being pedantic here)

- See the `get_callback_url` related discussion above.

**Comment by jleclanche** [comment]
@pennersr PR is ready for a new review.

By the way, if you are unhappy with the last commit, the first 6 can land without it. In fact, all the commits are independent from one another except for "Allow overriding the region in GET parameters" which requires `get_callback_url` and `valid_regions`.

**Comment by pennersr** [comment]
Thanks, LGTM

**Comment by pennersr** [comment]
... except for the pep8 error:

allauth/socialaccount/providers/battlenet/views.py:99:80: E501 line too long (80 > 79 characters)

**Comment by jleclanche** [comment]
@pennersr Fixed, but you really should change that - 80 characters is an unnecessarily low limit and it's making the code look bad in some places.

Also if flake8 issues are blocking, it should be set up on travis. I can submit a PR for that if you like.

**Comment by pennersr** [comment]
Thanks for the fix... (please don't file a PR on coding style... like pep8 as it is).

**Comment by jleclanche** [comment]
@pennersr I meant a PR to get flake8 set up on travis. If you're blocking PRs on it, it should be done automatically not at review time.

**Comment by pennersr** [comment]
That is already the case... the travis build did fail:

```
script:
  - flake8 --exclude=migrations allauth
  - coverage run manage.py test allauth
```
